noxfile.py

```diff
@@ -40,19 +40,66 @@ def set_environment_variables(env_dict, session):
 def run_pybamm_requires(session):
     """Download, compile, and install the build-time requirements for Linux and macOS: the SuiteSparse and SUNDIALS libraries."""  # noqa: E501
     set_environment_variables(PYBAMM_ENV, session=session)
+    force_rebuild = "--force" in session.posargs
+    SUNDIALS_LIBS = [
+        "libsundials_idas",
+        "libsundials_sunlinsolklu",
+        "libsundials_sunlinsoldense",
+        "libsundials_sunlinsolspbcgs",
+        "libsundials_sunlinsollapackdense",
+        "libsundials_sunmatrixsparse",
+        "libsundials_nvecserial",
+        "libsundials_nvecopenmp"
+    ]
+
+    KLU_LIBS = [
+        "libsuitesparseconfig",
+        "libklu",
+        "libamd",
+        "libcolamd",
+        "libbtf"
+    ]
+
+    if sys.platform == "linux":
+        sundials_so_path = [
+            Path(PYBAMM_ENV["LD_LIBRARY_PATH"], lib)
+            for lib in SUNDIALS_LIBS
+        ]
+        klu_so_path = [
+            Path(PYBAMM_ENV["LD_LIBRARY_PATH"], lib)
+            for lib in KLU_LIBS
+        ]
+
+    elif sys.platform == "darwin":
+        sundials_so_path = [
+            Path(PYBAMM_ENV["LD_LIBRARY_PATH"], lib + ".dylib")
+            for lib in SUNDIALS_LIBS
+    ]
+        klu_so_path = [
+            Path(PYBAMM_ENV["LD_LIBRARY_PATH"], lib + ".dylib")
+            for lib in KLU_LIBS
+    ]
+
```

Comment:
I tested this locally: here, `Path()` is just constructing PathLike objects and then you're doing a list comprehension over them, this approach doesn't check if the paths to these files actually exist. The correct flow should be to create paths from SUNDIALS_LIBS and KLU_LIBS and have one single function return a boolean if all of them are found. This value can be checked upon calling the function (whether it's `True` or `False`) and the session.install and session.run statements can be executed or skipped accordingly.

Reply:
```
SUNDIALS_LIBS = [
    # libs
    ]

KLU_LIBS = [
    # libs
    ]
    

def check_lib_paths(lib_names, lib_dir):

  if sys.platform == "darwin":
    ext = ".dylib"
  elif sys.platform == "linux":
    ext = ".so"

  paths = [Path(lib_dir, lib + ext) for lib in lib_names]  
  return all(os.path.exists(path) for path in paths)


@nox.session(name="pybamm-requires")
def run_pybamm_requires(session):
    # Set Variables
    
    sundials_path = check_lib_paths(SUNDIALS_LIBS, PYBAMM_ENV["LD_LIBRARY_PATH"])
    klu_path = check_lib_paths(KLU_LIBS, PYBAMM_ENV["LD_LIBRARY_PATH"])

    if sys.platform != "win32":
        session.install("wget", "cmake", silent=False)

        if sundials_path and klu_path and not force_rebuild:
           # Rest of the code
```
        
        
Something like this?

Reply:
When I test this locally, it doesn't exits early even if the dependencies are still present but if I change 

`if sundials_path and klu_path and not force_rebuild:` to `if sundials_path !=[ ] and klu_path != [ ] and not force_rebuild:`

it is exiting early

Reply:
You can modify `check_lib_paths` to iterate over the list of file paths and return either `True` or `False`, and change the condition in that case, which can then look like

```python
if check_lib_paths() and not force_rebuild:
    session.run("...")
```

i.e., the boolean `and` operator can then take care of the condition